### ISSUE TITLE
Error Report Duplications

### ISSUE THREAD

**Issue by DanVanAtta** [triplea-game/triplea#8230]
Also, this (#8180) is a duplicate of https://github.com/triplea-game/triplea/issues/7977.

In such cases, should the older or the newer be closed? I would say the newer, thus I would close this issue as duplicate.

Should we close this issue and continue this discussion in "7977"?

_Originally posted by @Cernelius in https://github.com/triplea-game/triplea/issues/8180#issuecomment-730883987_

**Comment by DanVanAtta** [comment]
@Cernelius   For 'Error Report' labelled items, use the latest issue. 'Error Report' refer to the items that are sent to us from the game-engine from the in-game error report feature.

The engine will already de-dupe the same problem across the same engine version. As a matter of course, we will summarily close all 'Error Report' items for older versions. For this reason, keep the latest open.

**Comment by Cernelius** [comment]
@DanVanAtta Thanks for the clarification, but I'm still unconvinced keeping the newest one open is preferable over keeping the oldest one open. I guess you are not sure, either, since I see this is labelled as a "discussion".

While the newest one gives a higher chance the problem is still actual, the oldest one gets us closer to the point whence it started (I guess ideal would be having the reference to the very first TripleA version presenting this problem, which, at least for regressions, is what would happen if having a near-to infinite pool of always reporting users).

For example, if one such error is opened and we start discussing it in the issue, then in the middle of it a new one is opened, we would close the one with the ongoing discussion in favour of the new one, then go there continuing the discussion and, maybe, suddenly another duplicate is created, so we have to migrate to the new one, since the previous one would get closed, and so on and so forth, splitting the discussion over an unlimited number of different issues (hopefully not restarting it each time). I get that the automatic de-dupe will limit this to different TripleA versions, thus moderating this process.

Moreover, keeping the old one would be consistent with how reported problems would be handled, as, if someone opens an issue about an existing problem, I assume one would tell him that this problem is already being tracked and close the new issue.

---

Everyone else agree that keeping closing existing duplicates as soon as a new one (meaning a duplicate made using a newer TripleA version) is made is the preferable process?

**Comment by DanVanAtta** [comment]
Well, 'discussion' does not mean 'open question' necessarily ;)
The labels are to help us filter out the topics, the queue of issues has grown quite considerably.

> While the newest one gives a higher chance the problem is still actual, the oldest one gets us closer to the point whence it started (I guess ideal would be having the reference to the very first TripleA version presenting this problem, which, at least for regressions, is what would happen if having a near-to infinite pool of always reporting users).

A new 'error report' against the latest version means the problem is current. Keep in mind that the backend will not recreate bug reports if there is one open for the current version. Hence, if we bump the version, all bug reports can then be created once again. This means any issues we fixed should not re-appear. Then, if we have a error report for the latest version, that means it's an ongoing problem, while a closed error report for the latest version means it was recently fixed and the fix is in the prerelease.

With respect to regressions, I think most interesting ones will not present as an error (game crash). Most interesting regressions are things like unit movement rules being wrong, various technology no longer working, etc..

While knowing when a problem is introduced is generally useful, we have not had bug reports forever so it's hard to know if something has always been around. For items that are regression in a prerelease, it's not too challenging to track them down. For things that have been ongoing from release to release, that are stack traces, it's more difficult to track down and often once we have a repro case and the stack trace then we can likely arrive at a fix.

> For example, if one such error is opened and we start discussing it in the issue, then in the middle of it a new one is opened, we would close the one with the ongoing discussion in favour of the new one, then go there continuing the discussion and, maybe, suddenly another duplicate is created, so we have to migrate to the new one, since the previous one would get closed, and so on and so forth, splitting the discussion over an unlimited number of different issues (hopefully not restarting it each time). I get that the automatic de-dupe will limit this to different TripleA versions, thus moderating this process.

Generally I would agree to favor the old report. The game-created bug reports are different though. Consider for example, we close all the 2.3 bug reports and that contains an original one. We then get the same bug report in 2.5, and close that one referencing the "closed" 2.3 bug report. At that point we have an open problem and all closed bug reports. The thing to keep in mind is once we have a new version we'll generally go back and close older bug reports (which is something todo right now, all of the 2.4 and 2.3 bug reports should be closed).

I'm being loose in calling these bug reports, specifically I'm referring to 'error reports' from in-game, which contain a stack trace for us to look at. Bug reports manually created are different. Most error reports should not have that much conversation in them other than "this is interesting", or "I wonder how this happened", and generally the conversation should not be too profound until we get to a fix.  With that said, I think our bug reports often contain too much conversation even and often need to be re-created as a task to summarize the information and allow a developer to skip most of the background reading and can just execute on fixing the task. A rule of thumb, if any issue gets over 3 pages long, and is not fixed within a week or two, it probably will never be fixed. That is just considering how much effort it takes someone fresh to simply *read* the dang thing. IF the amount of time to read an issue is a solid 45 minutes, it's likely more time than they had in the one sitting and they'll never even get started. Hence - long threads for bug reports are not good things, that is why I emphasize such stagnated problems need to be re-summarized as a task. Essentially re-created as:
- here is the repro case
- here is the bad behavior
- here is what the application should do instead

> Moreover, keeping the old one would be consistent with how reported problems would be handled, as, if someone opens an issue about an existing problem, I assume one would tell him that this problem is already being tracked and close the new issue.

Again, need to keep in mind that the auto-reports from in-game are different. It is a one-way street usually, most players do not seem to track the reports they create nor respond to comments in them. So players creating those reports literally never see them. The game does match error report title and game version and will tell them if one exists for that current version. The thinking is then on a new version we should not see the same error report or else it's something that was not fixed or still an issue (in which case a new one is created).

**Comment by DanVanAtta** [comment]
TL;DR
- error reports created from in-game are different from a bug report that is manually created. The former we prefer newest, the latter we prefer oldest
- all error reports of previous versions are closed summarily on new releases. If we favor old versions, then we'll be closing active problems to reference something that is closed. At that point it'll be unclear what has been fixed

**Comment by CharlieAtlas** [comment]
First off, besides where they came from, what do you consider the difference between and error report and a but report?  To me they seem like mostly the same thing besides the source.  

My thought would be to close the older one unless there is any relevant discussion, at which point the new one should point to it and be closed.  Otherwise just keep the newest one.

**Comment by DanVanAtta** [comment]
Error reports are anonymous and often lack any additional feedback once created. It would be great if the people that created them would track them closely, but that has not seemed to have been the case. Error reports are also closed when we get new versions, defect (bug) reports are closed when fixed.

I think the condition for closure is really the key distinction. Let's say you create a defect report, others will then keep track of that defect report, perhaps in forums. If we get a duplicate, and close the old one, then everyone's links become out of date. Let's say that process repeats, now you have a chain of issues to follow to get to the latest, and the conversation between all of them becomes scattered.

For error reports, if we kept the oldest, then when a new version of the game creates the same report, then that oldest item will already have been closed because it's for an old version. That is just a deal breaker, we need closed issues to represent items that are effectively closed and from then on only be historical reference where you can seek more context. Another point, nobody is likely to be keeping links to error reports, we shouldn't have this daisy-chain situation with them, only to know that we have a duplicate in the same version of the software and we will be working to resolve that problem one way or another.

**Comment by CharlieAtlas** [comment]
I guess i am not sure why you decide to close error reports just because they are old.  Is there just a hope that someone fixed them, but the issue didn't get closed?

**Comment by DanVanAtta** [comment]
Some bug reports also get closed if they are too old. The theory there is the length of the bug queue creates an overhead in of itself that makes the queue become longer. For example, spend 2 hours going through the queue, and you've now spent 2 hours going through a queue instead of fixing anything on that queue. A long queue drowns out other items. When it comes to bug reports, things that are a problem tend to come up again as a duplicate. Hence, there can be a cut-off where we admit defeat on some problems and just close them if they are really old, the idea being that the passing of N+1 days is not going to increase the odds that we go back and actually fix that problem.

For error reports we're being inundated with them, we don't need the same error reports across version 2.3, 2.4, and so on. The idea is when we get a new release, the old release is not going to be played anymore and anything that is still an issue will come up again because the server de-dupes by version. If something was truly fixed in 2.4, then we shouldn't see it in 2.5. Hence, 2.3 had a few hundred error reports, 2.4 had fewer, and with a general trend we shouldn't see too many error reports going forward. Closing error reports for older versions is an easy way to solve the duplication problem between versions, and helps us know we are working on something that is a current problem.

**Comment by DanVanAtta** [comment]
More directly, it's less hoping the problems got fixed, it's letting the current problems crop up again with fresh stack traces. It's an easy way to know we are working on active issues and not something that maybe accidentally was fixed. There are enough error reports there's very little  reason for someone to pick something to work on that was cut against an older version vs a recent one.
